src/main/java/com/axis/system/jenkins/plugins/downstream/cache/BuildCache.java

```diff
@@ -182,46 +209,82 @@ protected void updateCache(Run downstreamRun) {
   }
 
   public void doGarbageCollect() {
+    Logger.info("Running GC...");
     try (ACLContext ignored = ACL.as(ACL.SYSTEM)) {
       downstreamBuildCache
           .entrySet()
           .removeIf(
               e -> {
-                if (Run.fromExternalizableId(e.getKey()) == null) {
-                  LOGGER.info(e.getKey() + " will be GC:ed");
+                // Stop looking up builds if we are trying to shutdown.
+                if (!workerThreadPool.isShutdown()
```

Comment:
Why not throw an exception here instead and catch the exception outside. That way we will terminate faster. As it is now we will have to wait for the loop to finish. Not sure how long time that takes but I can imagine it is a while for a larger instance?

Reply:
Good point. Was not sure about the best way to exit a stream. Though looping through the build references in the cache map should not take even a second for even 1 million builds or so.

Reply:
After thinking about it, I "unstreamed" the loop to be able to easier control the flow.